PEP 562 __getattr__ for modules

pmp-p opened the pull request:
`__getattr__` for modules can help build a lazy loading / code unloading at runtime, load/unload one time use functions or LRU cache can help save a lot of ram.

the code used is a duplicate of https://github.com/micropython/micropython/blob/338635ccc64204b6f388cfaafca00e120090c622/py/objtype.c#L642

ref:

 https://www.python.org/dev/peps/pep-0562/

https://snarky.ca/lazy-importing-in-python-3-7/

pmp-p committed:
PEP 562 __getattr__ for modules

__getattr__ for modules can help build a lazy loading / code unloading at runtime, load/unload one time use functions or LRU cache can help save a lot of ram.

pfalcon reviewed (commented):
IMHO, implementing `__setattr__` is more useful/important, and thus should be done first: https://github.com/micropython/micropython/issues/3141

pmp-p commented on py/objmodule.c:
```diff
@@ -61,7 +61,20 @@ STATIC void module_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
         mp_map_elem_t *elem = mp_map_lookup(&self->globals->map, MP_OBJ_NEW_QSTR(attr), MP_MAP_LOOKUP);
         if (elem != NULL) {
             dest[0] = elem->value;
+        #ifndef MICROPY_CPYTHON_COMPAT
```
maybe #if MICROPY_PEP562, though incomplete because it should covers __dir__ too

setattr already works except for C/C++ modules which is normal behaviour afaik.

pmp-p committed:
Update objmodule.c

dpgeorge commented on py/objmodule.c:
```diff
@@ -61,7 +61,20 @@ STATIC void module_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+        } else {
+            if (attr != MP_QSTR___getattr__) {
+                mp_obj_t dest2[3];
+                mp_load_method_maybe(self_in, MP_QSTR___getattr__, dest2);
```
This could be optimised to use the existing `dest` array instead of a new `dest2`.  We know that if the call succeeds it will be a function, not a method, so only need 2 entries in dest.

dpgeorge replied:
Furthermore, it could use `mp_map_lookup(&self->globals->map, MP_OBJ_NEW_QSTR(MP_QSTR___getattr__), MP_MAP_LOOKUP)` directly.

dpgeorge commented on py/objmodule.c:
```diff
@@ -61,7 +61,20 @@ STATIC void module_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+                mp_load_method_maybe(self_in, MP_QSTR___getattr__, dest2);
+                if (dest2[0] != MP_OBJ_NULL) {
+                    dest2[2] = MP_OBJ_NEW_QSTR(attr);
+                    dest[0] = mp_call_method_n_kw(1, 0, dest2);
```
With the above optimisation, this would be something like `dest[0] = mp_call_function_2(elem->value, MP_OBJ_NEW_QSTR(attr))`.

dpgeorge reviewed (commented):
Thanks for the contribution.  It's a Python 3.7 feature but makes sense to add it.

Probably the config option should not reference PEP562 directly but rather the feature it adds, like `MICROPY_MODULE_SPECIAL_METHODS`

pmp-p committed:
changes following suggestions

tested on unix port, looks fine to me.

pmp-p committed:
oops style

pfalcon commented:
> but rather the feature it adds, like MICROPY_MODULE_SPECIAL_METHODS

Looking thru py/mpconfig.h, it doesn't take long to find MICROPY_MODULE_BUILTIN_INIT. It refers to builtin modules, so a natural counterpart for a feature proposed here will be MICROPY_MODULE_GETATTR.

pmp-p committed:
configuration tweaks

allow picking just __getattr__  with MICROPY_MODULE_GETATTR but keep place for a full implementation of modules specials including __getattr__ but not limited to.

dpgeorge commented:
@pmp-p are you able to add a test for this feature?  The way to do that would be to enable it on the unix coverage port (see mpconfigport_coverage.h) and add a test to the directory `tests/import/`

pmp-p committed:
add __getattr__ as a module special method

https://github.com/micropython/micropython/pull/4226

pmp-p committed:
test for __getattr__ on module

__getattr__ on module is added via MICROPY_MODULE_SPECIAL_METHODS or MICROPY_MODULE_GETATTR

pmp-p committed:
wrong folder

pmp-p committed:
test for __getattr_

__getattr__ on module (PEP562 /py3.7+ )  is activated by either MICROPY_MODULE_SPECIAL_METHODS or MICROPY_MODULE_GETATTR

dpgeorge commented:
Thanks @pmp-p for updating and adding the test.

But I see that you are using an anonymous email address for the commit.  We do require that contributions come with a valid email address, so there is some level of traceability of the code.  Are you willing to submit with a valid email address (not necessarily a personal one, but at least one which works and reaches you)?

pmp-p committed:
update mail for web git

pmp-p commented:
@dpgeorge  I un-checked anonymous web git operation but not sure about that ...

dpgeorge commented:
Ok, thanks for adding the email address.  This PR was squashed and merged in 454cca6016afc96deb6d1ad5d1b3553ab9ad18dd with some formatting changes, and updates to the new test so it is skippable when the getattr feature is not supported by a build.
